### PR TITLE
Circle CI: Run in Ubuntu and macOS environments, in parallel

### PR DESCRIPTION
**Pull request by janpio**
This PR takes the current Circle CI configuration for the macOS test run and creates a copy for Ubuntu, makes all the necessary changes (some commands have to be replaced, a dependency that is always present on macOS has to be manually installed, the caches need better names) and then adds a "workflow" that makes both test, macOS and Ubuntu, run at the same time.

### CODE REVIEW

**Comment by fastlane-bot-helper** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Pull Request is Work in Progress</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by janpio** [comment]
Success on first try:

image

👍

**Comment by janpio** [comment]
Although Github UI still shows this:
image
Bug or did I miss something?

**Comment by janpio** [comment]
Found something, seems to be a limitation of Github/Circle integration:
https://circleci.com/docs/2.0/workflows/#workflows-waiting-for-status-in-github
We can change that when this PR is merged.

**Comment by KrauseFx** [comment]
Wow 🚀

**Comment by janpio** [comment]
PR is now ready for review.

**Comment by KrauseFx** [comment]
screenshot 2017-12-19 08 56 43

Looks like CI is green, but the overall status is missing or something?

**Comment by KrauseFx** [comment]
This looks awesome, and I'm really excited to have Linux CI running also. Since this is a very Circle heavy PR, I'd love to get @DanToml's input on this before we merge. I know she's on vacation right now, so if it's cool from your side, let's wait for her to come back and get her input?

**Comment by janpio** [comment]
Absolutely, looking forward to it - this was my first contact with Circle CI.

Further work that benefits from Ubuntu or cross platform testing can just use this branch as base instead of `master`, so no harm done.

**Comment by janpio** [comment]
Oh, missed your second comment with the screenshot: Yes, that what I talked about in https://github.com/fastlane/fastlane/pull/11242#issuecomment-352146780 - you (or some other admin) would have to change a setting here in Github for this to be correct.

**Comment by KrauseFx** [comment]
Yeah, seems like it's something we should wait for until after the merge

**Comment by janpio** [comment]
Yes, as we know how to solve it (and CircleCI is aware and has this explicit instructions) I don't think this will become a problem to solve, we just shouldn't forget to do it when we merged this.

File: .circleci/config.yml
Comment (lacostej) [inline .circleci/config.yml]: We might want to provide a pre-built binary to avoid having to do that every time and maybe host it here: https://launchpad.net/ubuntu/+ppas

You could of course host it on github first.

File: .circleci/config.yml
Comment (lacostej) [inline .circleci/config.yml]: does /Users/distiller work on Linux? (store_test_results section)

Comment (janpio) [reply .circleci/config.yml]: afaik it doesn't even work on macOS - or I couldn't find out how it should work, what it should do etc. Do you have more insight?

File: .circleci/config.yml
Comment (janpio) [reply .circleci/config.yml]: That sounds like a good idea. I will look into that.

Comment (janpio) [reply .circleci/config.yml]: Fascinating, there are actually 4 sources for ready built `xar`s there I didn't find googling before:
https://launchpad.net/ubuntu/+ppas?name_filter=xar
Unfortunately none of them are v1.6.1, so I will see if I can build this myself.

Comment (janpio) [reply .circleci/config.yml]: https://launchpad.net/ubuntu/+ppas links to https://help.launchpad.net/Packaging/PPA links to https://help.launchpad.net/Packaging/PPA/Uploading links to https://help.launchpad.net/Packaging/PPA/BuildingASourcePackage links to http://packaging.ubuntu.com/html/ links to http://packaging.ubuntu.com/html/packaging-new-software.html - and is a bit complicated.

I think I will move this into a new issue and work on it independently from this PR. 

Do you think deploying it like this would be fine for the beginning @lacostej ?

Comment (janpio) [reply .circleci/config.yml]: https://github.com/fastlane/fastlane/issues/11280 now exists.

File: .circleci/config.yml
Comment (DanToml) [inline .circleci/config.yml]: no need to wrap these in quotes

File: .circleci/config.yml
Comment (DanToml) [inline .circleci/config.yml]: `v1-macos-gems-{{ checksum "Gemfile" }}` would be a better cache key here, because we need to be able to invalidate the cache key easily

File: .circleci/config.yml
Comment (DanToml) [inline .circleci/config.yml]: `v1-ubuntu-gems-{{ checksum "Gemfile" }}` would be a better cache key here, because we need to be able to invalidate the cache key easily

File: .circleci/config.yml
Comment (DanToml) [reply .circleci/config.yml]: We should build a docker container that includes these (based on the circleci/ruby:2.3 container)

File: .circleci/config.yml
Comment (DanToml) [reply .circleci/config.yml]: /Users/distiller will not work on linux

File: .circleci/config.yml
Comment (DanToml) [inline .circleci/config.yml]: No need to write .ruby-version on linux

File: .circleci/config.yml
Comment (DanToml) [inline .circleci/config.yml]: No need to update system rubygems on linux either, the circle image builds nightly

File: .circleci/config.yml
Comment (DanToml) [inline .circleci/config.yml]: Another candidate for the build docker container.
Ideally we should seperate shellcheck linting out and run it in its own job.

File: .circleci/config.yml
Comment (DanToml) [inline .circleci/config.yml]: Probably don't need a login shell on linux?

**Review by endocrimes — CHANGES_REQUESTED** [review]
This is a good start!

File: .circleci/config.yml
Comment (janpio) [reply .circleci/config.yml]: I wanted to indicate that these are chosen strings vs. stuff that from a list Circle expects. Bad idea?

File: .circleci/config.yml
Comment (janpio) [reply .circleci/config.yml]: Will look into what this actually does/means. Stupid copy/paste from macOS for now.

File: .circleci/config.yml
Comment (janpio) [reply .circleci/config.yml]: Got it.

File: .circleci/config.yml
Comment (janpio) [reply .circleci/config.yml]: Got it.

File: .circleci/config.yml
Comment (janpio) [reply .circleci/config.yml]: `v1` refers to what here? Is this v1 of your cache file? I was unclear if this references Circle CI 1.0 maybe somehow.

**Comment by KrauseFx** [comment]
Thanks for your help here @DanToml ❤️

File: .circleci/config.yml
Comment (janpio) [reply .circleci/config.yml]: (Why do we need this on macOS?)

Comment (janpio) [reply .circleci/config.yml]: Read up on it:
Don't think we need that.

https://circleci.com/docs/2.0/configuration-reference/#default-shell-options


(Why do we need this on macOS?)

Comment (janpio) [reply .circleci/config.yml]: Read up on it:
Don't think we need that.

https://circleci.com/docs/2.0/configuration-reference/#default-shell-options

(Why do we need this on macOS?)

Comment (janpio) [reply .circleci/config.yml]: Read up on it: https://www.gnu.org/software/bash/manual/html_node/Bash-Startup-Files.html#Bash-Startup-Files
Don't think we need that.

(Why do we need this on macOS?)

When `--login` is gone, all that is left is actually the default: https://circleci.com/docs/2.0/configuration-reference/#default-shell-options
So this line can be completely removed 👍

File: .circleci/config.yml
Comment (janpio) [reply .circleci/config.yml]: Found it: https://circleci.com/docs/2.0/caching/#clearing-cache

File: .circleci/config.yml
Comment (janpio) [reply .circleci/config.yml]: Ok, looked into this a bit:

This step is currently failing on our normal builds:
https://circleci.com/gh/fastlane/fastlane/12636
```
Unable to save test results from /Users/distiller/test-results
Error stat /Users/distiller/test-results: no such file or directory
Found no path with test results, skipping
```

But test results should actually be written to somewhere: https://github.com/fastlane/fastlane/blob/3830c51f153875ec0b7350af359709e32616cfab/Rakefile#L18

I will create an issue for this first, when it is fixed the result can be incorporated here.

Comment (janpio) [reply .circleci/config.yml]: This is now #11303 and PR #11304

File: .circleci/config.yml
Comment (janpio) [reply .circleci/config.yml]: See #11302

File: .circleci/config.yml
Comment (janpio) [reply .circleci/config.yml]: See #11302

File: .circleci/config.yml
Comment (janpio) [reply .circleci/config.yml]: PR #11304 is ready for review (and merge). It fixes this step and will probably work unchanged on Ubuntu as well, so can be copied over.

File: .circleci/config.yml
Comment (janpio) [inline .circleci/config.yml]: replace with result from #11302

**Comment by janpio** [comment]
Update:
- Waiting for review and merge of #11302 
- Dependency stuff that should be solved via a Docker container was "outsourced" to #11302

Besides that I hope I covered all comments.

**Comment by janpio** [comment]
#11302 is merged, I will incorporate this here now.

**Comment by janpio** [comment]
Rebased, and changes from #11304 adapted for the second half of the file. Circle CI now correctly collects the test reports of rspec and display them:

> Your build ran 4541 tests in RSpec with 0 failures
Slowest test: fastlane_core.spec.build_watcher_spec FastlaneCore::BuildWatcher.wait_for_build_processing_to_be_complete returns specified build when strict watch is enabled (took 10.00 seconds).

I think this is now "good" again and all relevant, outstanding feedback was addressed. Can I get another look at it please?

(As said before, I think the dependency and Docker stuff will be handled in a follow up issue: #11302)

File: .circleci/config.yml
Comment (DanToml) [reply .circleci/config.yml]: They're both the same when they get parsed anyway :woman_shrugging:, as much as I wish YAML required quotes everywhere, stylistically, I'm not a fan of them when they're not required, as it makes me assume something weird is going on.

File: .circleci/config.yml
Comment (DanToml) [reply .circleci/config.yml]: Yeah - We don't have a good way to clear caches through the UI/API yet, so the hack is to bump a version string in the yaml

File: .circleci/config.yml
Comment (DanToml) [reply .circleci/config.yml]: We need it on macOS jobs so that we source the `chruby` init scripts, to use our chosen ruby, and not system ruby

File: .circleci/config.yml
Comment (DanToml) [reply .circleci/config.yml]: :+1:

File: .circleci/config.yml
Comment (DanToml) [inline .circleci/config.yml]: Add a newline at the end and this is good to go

**Review by endocrimes — APPROVED** [review]

**Comment by DanToml** [comment]
@KrauseFx On merge, this will also require changing the required status' for merging, as the old `ci/circleci` named status doesn't get reported in workflows (because we get a status per job)

**Comment by janpio** [comment]
Oh, I actually **can't** merge this before @KrauseFx changes that setting as this is one of the expected checks ;) 

See here for instructions: https://circleci.com/docs/2.0/workflows/#workflows-waiting-for-status-in-github

**Pull request merged**
